### ISSUE TITLE
proposal: implement components use the new hook

### ISSUE THREAD

**Issue by jquense** [formatjs/formatjs#1290]
### FEATURE REQUEST

Right now react-intl adds a bunch of unneeded components for each usage, b.c `withIntl` adds a wrapper as well as a context.Consumer to the component stack. The v3 prerelease is a good moment to bump the minimum react version up to 16.8 and the library can make use of it's own `useIntl` hook to flatten app component trees.

I think this is generally worth doing given how many react-intl components tend to get rendered in an app compared to others (lots of messages). This isn't automatically a performance concern for apps but it could be, and is avoidable now. 

we've already mostyl done this work in our fork: https://github.com/4Catalyzer/react-intl/blob/master/src/components/message.js and i'd be happy to upstream it if ya'll are open to it.

this would purely be an internal optimization, no public api changes or lose of features

**Comment by leoyli** [comment]
I don't think it is reasonable to aggressively bump to v16.8... I'm pretty sure many projects wasn't upgraded yet. Now v3 will have `useIntl` for people using hooks but it is optional, so the minimum just need a v16.3 of React.  But I think it is also important to think about it and to draw a road map. (PS. I often do not use the build in components from `react-intl`...)

**Comment by jquense** [comment]
I don't now that it makes sense to hold off on the bump considering it's a completely non-breaking change to go from react 16.3 to 16.8. v3 is _already_ making the much bigger jump to only supporting v16+ which is a breaking change. That's all to say i'm not sure who can bump from 15 -> 16 but cannot make the bump from 16.3 -> 16.8

**Comment by leoyli** [comment]
Then perhaps, it could be done in v3.1?  Just like React v16.8, it won't be a serious breaks for v3.1. v 3.0 can just as a foundation for what you've mentioned.  I think what you've said involving many refactors and bench-marking on performance. I see no one here really have a time to do it.

For me, the minimum is to turn on `StrictMode` for my project, and I see that is also the major reason why we need a v3.  I'm kind of hoping it get a stable release, and there are no other blockers to delay it.

Also, thinking about the performance, code-splitting for translated strings is much worth to be done, IMHO.

**Comment by DragonRaider5** [comment]
Tbh I don't think using hooks over the standard context consumers is a performance improvement after all.

Context consumers don't add to the rendered DOM and as such don't have much of a performance impact (at least compared to hooks) AFAIK.

**Comment by jquense** [comment]
It's not just the Consumers, its the added wrapper and consumer as well as the additional forwardRef. like i said tho,  it's not a strict perf issue. I'm not arguing that doing this would be a clear perf win, but in general More Components, is less good then Less Components in terms of time spent reconciling and rendering. In this instance there is almost no downside to using hooks.

> Then perhaps, it could be done in v3.1?

that would be a breaking a change for react-intl, since now you are requiring a high react version, that's a slightly different scenario. It would be free for a user to upgrade to that react, but it'd still require that they do so before their app worked again.

> and there are no other blockers to delay it.

sure, you could require the version now while it's still a prerelease even if not specifically required. otherwise what would be the plan for upgrading? quick v4 after that? 

This isn't really a big deal but its also very low cost now, and higher cost later, and the work is already done (in our fork)

**Comment by leoyli** [comment]
@jquense,

I personally think it would be really nice to remove all these `Customer` component that makes react devTool much cleaner.  But as far as I know, and mentioned, there are some relatively conservative team that are resist to upgrade to React 16.8 but they are upgraded to 16.3 already. Not all people like the shiny hooks, and there are new things to learn as an investment for some team. I'm not sure what is the attitude about the goal of v3 here, so I would ask @redonkulus, @DragonRaider5 to make further clarifications.

And since the work have been done in your fork, perhaps you can submit an PR so we can get it a review? I'm pretty sure sooner or later we will have to do this.

**Comment by DragonRaider5** [comment]
@jquense if it's not about performance, what is it about?
Do you feel like the wrapper messes up something?

**Comment by jquense** [comment]
It is about performance (as well as debuggability to a lesser extent). I've been hedging on that because "do less work" is a hard to quantify win, since it is very situation dependent. React Intl is a ubiquitous sort of library, the components it exports are _everywhere_ in an app, so I think it's particularly beneficial for the library to err on side of adding as little overhead as it can. 

I definitely understand the need to make sure you bring your users with you on upgrades, I maintain a lot of react libraries myself. In general tho v3 is _already_ asking users to make the hardest jump, the breaking change from 15 -> 16. Asking them to bump to the latest react version cost them nothing more after that. It also doesn't require that react-intl users like or accept hooks, this change is an internal one, it doesn't force a new API on users, it only potentially makes their apps a bit faster and easier to debug.

**Comment by DragonRaider5** [comment]
@jquense would you maybe like to add some performance tests in order to make us able to test if there is a performance benefit?

I think we might even be able to have both implementations alongside each other,  just not sure whether that's worth it.

**Comment by jquense** [comment]
There aren't really good ways to write performance tests for this sort of thing because the effect is specific to apps (not my position, but one the react core team notes often). It's sort of by definition a thing that doesn't test well. Maybe a different question is what is the down-side to doing less work, and easier to debug, while also modernizing the code? Does that downside warrant not making a change?

**Comment by longlho** [comment]
I'd prob wanna hold off the hooks train until it's had some more bake time. A lot of consumers aren't on 16.8. There're no strong reason to migrate to hooks as of right now.

**Comment by jquense** [comment]
sure, ya'll should do what you think is best for your user base. We've already done the work of forking so no ultimate skin off our back :)

 I would say as someone well established and active in the react ecosystem, hooks are stable and production ready, they were released b/c of extremely high confidence that they were "baked" so I would be a bit cautious of waiting for some stronger signal from the react core team that they are ready to use

**Comment by Grsmto** [comment]
Most "big" libraries already support hooks in their latest version. Actually when I saw that I needed a HoC to use the imperative APIs I immediately searched for "hooks" in the issue tracker.
It's a big +1 for me as most users that will upgrade to latest `react-intl` will probably upgrade to latest `react` as well anyway. 16.8 has been released a while ago already.

**Comment by longlho** [comment]
I think it's fair to say we'll convert to hooks internally once enzyme officially supports it (https://github.com/airbnb/enzyme/issues/1553 for tracking purposes), otherwise it'd break existing tests.

Converting to `static contextType` has broken enzyme tests internally within DBX so I can't even switch to that on `class`-based components.

**Comment by jquense** [comment]
What enzyme issues are you running into? I've got a lot of react libraries that we all test with enzyme and hooks with no issues. including this, which is entirely hooks: https://github.com/react-restart/hooks

**Comment by jquense** [comment]
O this is a shallow rendering issue? That problem is not likely to be solved any time soon, because it's basically intended. React's shallow renderer does not support Context, and while there is a PR up it's been sitting with no attention from the core team for a year.

In general I think it's considered better to not use shallow rendering anymore anyway. It's not much simpler or faster than normal rendering. Would you be open to having the tests converted so that user facing features aren't blocked by test utilities?

**Comment by longlho** [comment]
Yup it is the shallow rendering issue. Can you link the PR you’re
mentioning for tracking purposes? The problem I’m having w/ mount is that
deeply-nested components can depend on upstream Provider that the testing
code is unaware of. This becomes very hard to scale in a large eng org if
all Providers have to be specified.

**Comment by jquense** [comment]
It's definitely harder in a larder app setting I'd agree. For my work we have a helper that mounts and mocks the right mock contexts for the things that are present everywhere, e.g. intl, router, theme, etc. For more focused stuff we tend to do something like: https://github.com/4Catalyzer/react-intl/blob/master/test/unit/testUtils.js\

you can also use jest.mock to have the Context object itself return with a default value, or structure the contexts to have a noop default value anyway.

upstream, upstream PR: https://github.com/facebook/react/pull/14329 

I'm honestly just jumping to be neighborly, i've done a bunch of v16 work on enzyme and react around the testing, and while i'm _definitely_ not an authority, my reading of the situation over the last ~1.5years is that the shallow renderer is a bit of a thorn, b/c it's not clear how some of these new features should/could be implemented in that context, and it's not a big priority to figure it out. The happier path right now is rendering normally, even with all the additional frustrations there.

**Comment by longlho** [comment]
Thanks a lot for all the info. I've surfaced it internally within DBX and luckily there aren't a huge amount of `shallow` usage, but quite a few to migrate away from. I'll keep this issue open until we can move to 16.8 fully.
